pwnlib/protocols/adb/__init__.py

```diff
@@ -33,17 +34,19 @@ def pack(val):
 def unpack(val):
     return int(val, 16)
 
-OKAY = "OKAY"
-FAIL = "FAIL"
+OKAY = b"OKAY"
+FAIL = b"FAIL"
 
 class Message(object):
     """An ADB hex-length-prefixed message"""
     def __init__(self, string):
         self.string = string
+    def __bytes__(self):
+        return self.__flat__()
     def __str__(self):
-        return ('%04x' % len(self.string)) + self.string
+        return self.__flat__()
```

Comment:
For Python3, I expect that `__flat__()` MUST return a `bytes` object.

I don't see how this can be a valid return type for `str()`.

Reply:
This is working for python 2 (`bytes(obj)` calls `obj.__str__()`) and not covered at all for python 3 (`bytes(obj)` calls `obj.__bytes__()`). While this looks not like the best solution, the protocol objects are not exported anywhere near the user, so they should not experience any issues.

Reply:
I think the proper approach here then might be to ONLY provide `__flat__`.  Do we invoke `str()` anywhere?

Separately, we might want to consider dropping `__flat__` Pwntools-wide and instead prefer `__bytes__`.  This will ensure the "right" thing happens on Python3.

Another hack that we might make is just to monkey-patch the toplevel `bytes()` method on Python2.  I don't expect it's invoked very much (certainly not as much as `str()`), so I don't think the performance impact would be much.  

We could then invoke `__bytes__` on anything that has it.  The `flat()` family of functions can just be updated to invoke `__bytes__` instead.

Reply:
Email z wtorek, 27 listopada 2018 od Zach Riggle:
> I think the proper approach here then might be to ONLY provide `__flat__`. Do we invoke `str()` anywhere?

We don't, but I'm nearly sure we do invoke `bytes`, which is `str` under python 2.

> Separately, we might want to consider dropping `__flat__` Pwntools-wide and instead prefer `__bytes__`. This will ensure the "right" thing happens on Python3.

This could be an option, but I think we will then need to invent a class decorator like [`six.python2_unicode_compatible`](https://six.readthedocs.io/#six.python_2_unicode_compatible), say `python2_bytes_compatible`, which would act like:

```py
def python2_bytes_compatible(klass):
    if six.PY2:
        klass.__str__ = __bytes__
    return klass

# And then, for example

@python2_bytes_compatible
class ROPChain:
    # ...
```

> Another hack that we might make is just to monkey-patch the toplevel `bytes()` method on Python2. I don't expect it's invoked very much (certainly not as much as `str()`), so I don't think the performance impact would be much.

We would need to have in mind that it is used in instance checking calls like `isinstance(x, bytes)`, hence it cannot be a function nor some subclass; so it would be better to do as I wrote above.

> We could then invoke `__bytes__` on anything that has it. The `flat()` family of functions can just be updated to invoke `__bytes__` instead.

Actually, this could be a good idea (but I still support the idea of introducing `__flat_at__` or similar).